Hypercore API ref

billy-the-fish opened the pull request.

billy-the-fish committed:
feat: update structure for hypercore

billy-the-fish committed:
chore: space to validate initial build.

billy-the-fish committed:
chore: add first hypercore files.

billy-the-fish committed:
chore: add api ref files and alter_table to the build.

billy-the-fish committed:
chore: add remove_columnstore_policy to the build.

billy-the-fish committed:
chore: add convert_to_columnstore to the build.

billy-the-fish committed:
chore: add convert_to_rowstore to the build.

billy-the-fish committed:
chore: updates for build.

billy-the-fish committed:
chore: updates for build.

billy-the-fish committed:
chore: remove links from steps in a list.

billy-the-fish committed:
chore: remove links from steps in a list.

billy-the-fish committed:
Merge branch 'latest' into 235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore

billy-the-fish committed:
chore: hopefully this will work now.

billy-the-fish committed:
chore: add deprecation for compression.

billy-the-fish committed:
chore: simplify workflow, keep code for main docs.

billy-the-fish committed:
chore: updates on review. .

billy-the-fish committed:
Merge branch 'latest' into 235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore

billy-the-fish committed:
chore: updates on review.

billy-the-fish committed:
Merge branch 'latest' into 235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore

billy-the-fish committed:
Merge branch 'latest' into 235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore

philkra commented:
@billy-the-fish the changes look good to me. One thing, I would look like to add is a reference from the deprecated functions / parameters to the new ones. For example:
[compress_chunk](https://docs-dev.timescale.com/docs-235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore/api/235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore/compression/compress_chunk/) is deprecated in favour of [convert_to_columnstore](https://docs-dev.timescale.com/docs-235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore/api/235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore/hypercore/convert_to_columnstore/). As user, I would love to know if there is a replacement for the deprecated function, having a reference to the replacement would be handy.

billy-the-fish committed:
chore: add replacedby link. Has to be hardcoded so will not work until these docs go live.

billy-the-fish committed:
Merge branch 'latest' into 235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore

erimatnor commented on _partials/_hypercore_policy_workflow.md:
````diff
@@ -0,0 +1,57 @@
+
+1. **Remove a columnstore policy**
+   ``` sql
+   SELECT remove_columnstore_policy('older_stock_prices');
````
I believe the columnstore API is now using procedures, which use the `CALL` command instead of `SELECT`
```suggestion
   CALL remove_columnstore_policy('older_stock_prices');
```

erimatnor commented on _partials/_hypercore_policy_workflow.md:
````diff
@@ -0,0 +1,57 @@
+
+   For example, 60 days after the data was added to the table:
+   ``` sql
+   SELECT add_columnstore_policy('older_stock_prices', after => INTERVAL '60d');
````
I believe the columnstore API is now using procedures, which use the CALL command instead of SELECT

```suggestion
   CALL add_columnstore_policy('older_stock_prices', after => INTERVAL '60d');
```

erimatnor commented on api/hypercore/add_columnstore_policy.md:
````diff
@@ -0,0 +1,108 @@
+
+   * 60 days after the data was added to the table:
+     ``` sql
+     SELECT add_columnstore_policy('stocks_real_time', after => INTERVAL '60d');
````
```suggestion
     CALL add_columnstore_policy('stocks_real_time', after => INTERVAL '60d');
```

erimatnor commented on api/hypercore/add_columnstore_policy.md:
````diff
@@ -0,0 +1,108 @@
+   * 3 months prior to the moment you run the query:
+
+     ``` sql
+     SELECT add_columnstore_policy('stocks_real_time', created_before => INTERVAL '3 months');
````
```suggestion
     CALL add_columnstore_policy('stocks_real_time', created_before => INTERVAL '3 months');
```

erimatnor commented on api/hypercore/add_columnstore_policy.md:
````diff
@@ -0,0 +1,108 @@
+   * With an integer-based time column:
+
+     ``` sql
+     SELECT add_columnstore_policy('table_with_bigint_time', BIGINT '600000');
````
```suggestion
     CALL add_columnstore_policy('table_with_bigint_time', BIGINT '600000');
```

erimatnor commented on api/hypercore/add_columnstore_policy.md:
````diff
@@ -0,0 +1,108 @@
+   * Older than eight weeks:
+
+     ``` sql
+     SELECT add_columnstore_policy('cpu_weekly', INTERVAL '8 weeks');
````
```suggestion
     CALL add_columnstore_policy('cpu_weekly', INTERVAL '8 weeks');
```

erimatnor commented on api/hypercore/convert_to_columnstore.md:
````diff
@@ -0,0 +1,53 @@
+To compress a single chunk:
+
+``` sql
+SELECT convert_to_columnstore('_timescaledb_internal._hyper_1_2_chunk');
````
```suggestion
CALL convert_to_columnstore('_timescaledb_internal._hyper_1_2_chunk');
```

erimatnor commented on api/hypercore/convert_to_columnstore.md:
```diff
@@ -0,0 +1,53 @@
+
+## Samples
+
+To compress a single chunk:
```
```suggestion
To convert a single chunk to columnstore:
```

erimatnor commented on api/hypercore/convert_to_rowstore.md:
````diff
@@ -0,0 +1,90 @@
+   - Convert single chunk:
+
+      ``` sql
+      SELECT convert_to_rowstore('_timescaledb_internal._hyper_2_2_chunk');
````
```suggestion
      CALL convert_to_rowstore('_timescaledb_internal._hyper_2_2_chunk');
```

erimatnor commented on api/hypercore/convert_to_rowstore.md:
````diff
@@ -0,0 +1,90 @@
+      SELECT convert_to_rowstore('_timescaledb_internal._hyper_2_2_chunk');
+      ```
+
+   - Convert all chunks in a hypertable named `metrics`:
+
+      ``` sql
+      SELECT convert_to_rowstore(c, true) FROM show_chunks('metrics') c;
+      ```
````
Since `convert_to_rowstore` is a procedure, this type of query is no longer supported. 

It is possible to write a wrapper procedure to convert all chunks, like this, but maybe we should just remove this from the docs for now? Normally, users rely on a policy to convert in any case.

billy-the-fish replied:
Updated, thank you.

erimatnor commented on api/hypercore/convert_to_rowstore.md:
````diff
@@ -0,0 +1,90 @@
+1. **Convert the updated chunks back to the columnstore**
+
+   ``` sql
+   SELECT convert_to_columnstore('_timescaledb_internal._hyper_1_2_chunk');
````
```suggestion
   CALL convert_to_columnstore('_timescaledb_internal._hyper_1_2_chunk');
```

erimatnor commented on api/hypercore/index.md:
````diff
@@ -0,0 +1,93 @@
+
+   For example, 60 days after the data was added to the table:
+   ``` sql
+   SELECT add_columnstore_policy('older_stock_prices', after => INTERVAL '60d');
````
```suggestion
   CALL add_columnstore_policy('older_stock_prices', after => INTERVAL '60d');
```

erimatnor commented on api/hypercore/page-index/page-index.js:
```diff
@@ -0,0 +1,55 @@
+      {
+        title: "chunk_columnstore_settings",
+        href: "chunk_columnstore_settings",
+        excerpt: "Show the compression settings for each chunk that has compression enabled",
```
```suggestion
        excerpt: "Show the columnstore settings for each chunk that is a columnstore",
```

erimatnor commented on api/hypercore/page-index/page-index.js:
```diff
@@ -0,0 +1,55 @@
+      {
+        title: "add_columnstore_policy",
+        href: "add_columnstore_policy",
+        excerpt: "Compresses a chunk automatically in the background after it reaches a given age",
```
```suggestion
        excerpt: Convert a chunk to columnstore automatically in the background after it reaches a given age",
```

erimatnor commented on api/hypercore/page-index/page-index.js:
```diff
@@ -0,0 +1,55 @@
+      {
+        title: "columnstore_settings",
+        href: "columnstore_settings",
+        excerpt: "Get information about compression-related settings for hypertables",
```
```suggestion
        excerpt: "Get information about columnstore-related settings for hypertables",
```

erimatnor commented on api/hypercore/page-index/page-index.js:
```diff
@@ -0,0 +1,55 @@
+      {
+        title: "chunk_columnstore_stats",
+        href: "chunk_columnstore_stats",
+        excerpt: "Get chunk-specific statistics related to hypertable compression",
```
```suggestion
        excerpt: "Get statistics for columnstore chunks",
```

erimatnor commented on api/hypercore/page-index/page-index.js:
```diff
@@ -0,0 +1,55 @@
+      {
+        title: "convert_to_columnstore",
+        href: "convert_to_columnstore",
+        excerpt: "Compress or recompress a specific chunk in the rowstore and add it to the columnstore",
```
```suggestion
        excerpt: "Convert a specific chunk from rowstore to columnstore",
```

erimatnor commented on api/hypercore/page-index/page-index.js:
```diff
@@ -0,0 +1,55 @@
+      {
+        title: "convert_to_rowstore",
+        href: "convert_to_rowstore",
+        excerpt: "Decompress a chunk from the columnstore and add it to the rowstore",
```
```suggestion
        excerpt: "Convert a specific chunk from columnstore to rowstore",
```

erimatnor commented on api/hypercore/page-index/page-index.js:
```diff
@@ -0,0 +1,55 @@
+      {
+        title: "hypertable_columnstore_settings",
+        href: "hypertable_columnstore_settings",
+        excerpt: "Returns information about the compression settings for each hypertable in the columnstore.",
```
```suggestion
        excerpt: "Returns information about the columnstore settings for each hypertable",
```

erimatnor commented on api/hypercore/page-index/page-index.js:
```diff
@@ -0,0 +1,55 @@
+      {
+        title: "hypertable_columnstore_stats",
+        href: "hypertable_columnstore_stats",
+        excerpt: "Get statistics related to hypertable compression",
```
```suggestion
        excerpt: "Get columnstore statistics for hypertables",
```

erimatnor commented on api/hypercore/page-index/page-index.js:
```diff
@@ -0,0 +1,55 @@
+      {
+        title: "remove_columnstore_policy",
+        href: "remove_columnstore_policy",
+        excerpt: "Remove the compression policy",
```
```suggestion
        excerpt: "Remove the columnstore policy",
```

erimatnor commented on api/hypercore/remove_columnstore_policy.md:
````diff
@@ -0,0 +1,47 @@
+- **Remove the columnstore policy from the `cpu` table**:
+
+   ``` sql
+   SELECT remove_columnstore_policy('cpu');
````
```suggestion
   CALL remove_columnstore_policy('cpu');
```

erimatnor commented on api/hypercore/remove_columnstore_policy.md:
````diff
@@ -0,0 +1,47 @@
+- **Remove the columnstore policy from the `cpu_weekly` continuous aggregate**:
+
+   ``` sql
+   SELECT remove_columnstore_policy('cpu_weekly');
````
```suggestion
   CALL remove_columnstore_policy('cpu_weekly');
```

erimatnor reviewed (commented):
Made a quick pass and added some suggestions.

billy-the-fish committed:
Apply suggestions from code review

Awesome review, thank you.

Co-authored-by: Erik Nordström <819732+erimatnor@users.noreply.github.com>
Signed-off-by: Iain Cox <iain@timescale.com>

billy-the-fish committed:
chore: update on review.

billy-the-fish committed:
chore: add enable_segmentwise_recompression.

billy-the-fish committed:
chore: add enable_segmentwise_recompression.

jonatas commented on api/compression/index.md:
```diff
@@ -30,6 +31,8 @@ Compressed chunks have the following limitations:
     after constraint creation.
 *   [Timescale SkipScan][skipscan] does not currently work on compressed chunks.
 
+<Deprecated2180 /> Replaced by <a href="https://docs.timescale.com/api/latest/hypercore/">Hypercore</a>.
```
We may emphasize that compression works in the new version.

billy-the-fish replied:
Done

billy-the-fish committed:
Merge branch 'latest' into 235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore

billy-the-fish committed:
Merge branch 'latest' into 235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore

billy-the-fish committed:
chore: update deprecation notices.

billy-the-fish committed:
chore: update deprecation notices.

mkindahl commented on api/hypercore/convert_to_columnstore.md:
```diff
@@ -0,0 +1,53 @@
+
+|Column| Type                                            | Description                                     |
+|-|-------------------------------------------------|-------------------------------------------------|
+| `convert_to_columnstore` | REGCLASS | The name of the chunk added to the columnstore. |
```
The result set here does not have a name, so not sure how you want to handle this. We have two different cases: 1) return a table-like result set with zero or more rows, and 2) return a single value of a specific type. This one just returns a `regclass` with no name.

billy-the-fish replied:
Hi @mkindahl , I've updated with my proposal of how to handle this case. Not entirely sure of the words to add so have left it open.

mkindahl commented on api/hypercore/convert_to_columnstore.md:
```diff
@@ -0,0 +1,53 @@
+|----------------------|--|---------|--|----------------------------------------------------------------------------------------------------------------|
+| `chunk`         | REGCLASS | -       |✔| Name of the chunk to add to the columnstore.                                                                   |
+| `if_not_columnstore` | BOOLEAN | `true`  |✖| Set to `false` so this job fails with an error rather than a warning if `chunk` is already in the columnstore. |
+| `recompress`         | BOOLEAN | `false` |✖| Set to `true` to add a chunk that had more data inserted after being added to the columnstore.                 |
```
This parameter requires a more elaborate description. How do we add that to the reference documentation? Do we have it after the table with the short description?

billy-the-fish replied:
Best would be to add a sample that contains the more elaborate description.

mkindahl replied:
But where? For this case, a simple example will not be good enough. We need to describe what recompression does compared to compression.

billy-the-fish replied:
Hi @mkindahl sorry for the late reply. Would you mind writing down what needs to be said and I will figure out how to add it?

mkindahl replied:
>If `recompress` is true, TimescaleDB will attempt to use an optimized method for recompressing data that have been partially compressed as a result of modifications to the chunk. This is usually more efficient, but in some cases it can result is a more expensive operation. Setting this parameter to `false` will result in a complete decompression of the data followed by a compression.

billy-the-fish replied:
Nice, updated.

billy-the-fish committed:
chore: remove stuff breaking the build.

mkindahl commented on api/hypercore/add_columnstore_policy.md:
```diff
@@ -0,0 +1,108 @@
+| `schedule_interval`       |INTERVAL| 12 hours when [chunk_time_interval][chunk_time_interval] >= `1 day` for `hypertable`. Otherwise `chunk_time_interval` / `2`. | ✖        | Set the interval between the finish time of the last execution of this policy and the next start.|
+| `initial_start`     |TIMESTAMPTZ| The interval from the finish time of the last execution to the [next_start][next-start].| ✖| Set the time this job is first run. This is also the time that `next_start` is calculated from.|
+| `timezone`          |TEXT| UTC. However, daylight savings time(DST) changes may shift this alignment. | ✖ | Set to a valid time zone to mitigate DST shifting. If `initial_start` is set, subsequent executions of this policy are aligned on `initial_start`.|
+| `if_not_exists`     |BOOLEAN| `false` | ✖ | Set to `true` so this job fails with a warning rather than an error if a columnstore policy already exists on `hypertable` |
```
This is not required
```suggestion
| `if_not_exists`     |BOOLEAN| `false` |  | Set to `true` so this job fails with a warning rather than an error if a columnstore policy already exists on `hypertable` |
```

billy-the-fish replied:
✔ indicated required, ✖  not required.

mkindahl replied:
Ah, I thought it was "crossed" and "not-crossed". Sorry about that.

mkindahl commented on api/hypercore/add_columnstore_policy.md:
```diff
@@ -0,0 +1,108 @@
+| `created_before` |INTERVAL| NULL | ✖ | Add chunks with a creation time of `now() - created_before` to the columnstore. <br/> `created_before` is <ul><li>Not supported for continuous aggregates.</li><li>Mutually exclusive with `after`.</li></ul> |
+| `schedule_interval`       |INTERVAL| 12 hours when [chunk_time_interval][chunk_time_interval] >= `1 day` for `hypertable`. Otherwise `chunk_time_interval` / `2`. | ✖        | Set the interval between the finish time of the last execution of this policy and the next start.|
+| `initial_start`     |TIMESTAMPTZ| The interval from the finish time of the last execution to the [next_start][next-start].| ✖| Set the time this job is first run. This is also the time that `next_start` is calculated from.|
+| `timezone`          |TEXT| UTC. However, daylight savings time(DST) changes may shift this alignment. | ✖ | Set to a valid time zone to mitigate DST shifting. If `initial_start` is set, subsequent executions of this policy are aligned on `initial_start`.|
```
This is not required:
```suggestion
| `timezone`          |TEXT| UTC. However, daylight savings time(DST) changes may shift this alignment. |  | Set to a valid time zone to mitigate DST shifting. If `initial_start` is set, subsequent executions of this policy are aligned on `initial_start`.|
```

mkindahl commented on api/hypercore/add_columnstore_policy.md:
```diff
@@ -0,0 +1,108 @@
+| `hypertable`              |REGCLASS| - | ✔ | Name of the hypertable or continuous aggregate to run this [job][job] on.|
+| `after`          |INTERVAL or INTEGER|- | ✖ | Add chunks containing data older than `now - {after}::interval` to the columnstore. <br/> Use an object type that matchs the time column type in `hypertable`: <ul><li><b><code>TIMESTAMP</code>, <code>TIMESTAMPTZ</code>, or <code>DATE</code></b>: use an <code>INTERVAL</code> type.</li><li><b> Integer-based timestamps </b>: set an integer type using t ...
+| `created_before` |INTERVAL| NULL | ✖ | Add chunks with a creation time of `now() - created_before` to the columnstore. <br/> `created_before` is <ul><li>Not supported for continuous aggregates.</li><li>Mutually exclusive with `after`.</li></ul> |
+| `schedule_interval`       |INTERVAL| 12 hours when [chunk_time_interval][chunk_time_interval] >= `1 day` for `hypertable`. Otherwise `chunk_time_interval` / `2`. | ✖        | Set the interval between the finish time of the last execution of this policy and the next start.|
```
This is not required:
```suggestion
| `schedule_interval`       |INTERVAL| 12 hours when [chunk_time_interval][chunk_time_interval] >= `1 day` for `hypertable`. Otherwise `chunk_time_interval` / `2`. |         | Set the interval between the finish time of the last execution of this policy and the next start.|
```

mkindahl reviewed (commented):
I noted that many entries are noted as required while they are not.

billy-the-fish committed:
chore: remove stuff breaking the build.

billy-the-fish committed:
Merge branch 'latest' into 235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore

billy-the-fish committed:
chore: updates on review.

billy-the-fish committed:
Merge branch 'latest' into 235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore

erimatnor committed:
Add API description for merge_chunks() (#3738)

TimescaleDB 2.18 includes a new procedure to merge chunks. This change
adds the API documentation.

billy-the-fish committed:
Merge branch 'latest' into 235-docs-rfc-update-the-api-ref-for-hyperstore-and-compression-for-any-changes-due-to-the-hyperstore

billy-the-fish committed:
chore: update return table.

billy-the-fish committed:
chore: change from deprecated to Old API.

billy-the-fish committed:
chore: change from deprecated to Old API.

billy-the-fish committed:
chore: update on review.
